From 5824f7a5a6d8dd68d54a7c26022ad7117a646479 Mon Sep 17 00:00:00 2001
From: Pat Noziska <patrick.j.noziska@intel.com>
Date: Thu, 14 Nov 2013 09:53:14 -0800
Subject: [PATCH 118/429] Fix Critical Klocwork issues

Fix Critical Klocwork issues in linux/kernel, mostly
potential NULL pointer derefernces and potential
buffer overflows.

Signed-off-by: Pat Noziska <patrick.j.noziska@intel.com>
---
 arch/x86/include/asm/apb_timer.h | 1 +
 drivers/acpi/osl.c               | 5 +++++
 fs/ecryptfs/crypto.c             | 4 +++-
 kernel/power/process.c           | 2 ++
 4 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/apb_timer.h b/arch/x86/include/asm/apb_timer.h
index 0acbac299e49..2a67ddac22a3 100644
--- a/arch/x86/include/asm/apb_timer.h
+++ b/arch/x86/include/asm/apb_timer.h
@@ -44,6 +44,7 @@ extern int sfi_mtimer_num;
 
 static inline unsigned long apbt_quick_calibrate(void) {return 0; }
 static inline void apbt_time_init(void) { }
+static inline void apbt_setup_secondary_clock(void) { }
 
 #endif
 #endif /* ASM_X86_APBT_H */
diff --git a/drivers/acpi/osl.c b/drivers/acpi/osl.c
index 11441ad69de3..2762f611df26 100644
--- a/drivers/acpi/osl.c
+++ b/drivers/acpi/osl.c
@@ -707,6 +707,11 @@ acpi_os_physical_table_override(struct acpi_table_header *existing_table,
 		table = acpi_os_map_memory(acpi_tables_addr + table_offset,
 					   ACPI_HEADER_SIZE);
 
+		if (table == NULL) {
+			pr_err("Error mapping ACPI memory\n");
+			return AE_ERROR;
+		}
+
 		if (table_offset + table->length > all_tables_size) {
 			acpi_os_unmap_memory(table, ACPI_HEADER_SIZE);
 			WARN_ON(1);
diff --git a/fs/ecryptfs/crypto.c b/fs/ecryptfs/crypto.c
index 1da2446bf6b0..3cf8b3a2ba3d 100644
--- a/fs/ecryptfs/crypto.c
+++ b/fs/ecryptfs/crypto.c
@@ -1023,8 +1023,10 @@ int ecryptfs_new_file_context(struct inode *ecryptfs_inode)
 		       "to the inode key sigs; rc = [%d]\n", rc);
 		goto out;
 	}
+	/* TODO: Investigate side effect of truncating name if too long */
 	cipher_name_len =
-		strlen(mount_crypt_stat->global_default_cipher_name);
+		min(strlen(mount_crypt_stat->global_default_cipher_name),
+		    sizeof(crypt_stat->cipher)-1);
 	memcpy(crypt_stat->cipher,
 	       mount_crypt_stat->global_default_cipher_name,
 	       cipher_name_len);
diff --git a/kernel/power/process.c b/kernel/power/process.c
index 0695319b5fde..a7bdeeab5605 100644
--- a/kernel/power/process.c
+++ b/kernel/power/process.c
@@ -33,6 +33,8 @@ static int try_to_freeze_tasks(bool user_only)
 	u64 elapsed_csecs64;
 	unsigned int elapsed_csecs;
 	bool wakeup = false;
+	int sleep_usecs = USEC_PER_MSEC;
+	char *busy_wq_name = NULL;
 
 	do_gettimeofday(&start);
 
-- 
2.37.3

